devlib/target_runner.py

```diff
@@ -0,0 +1,267 @@
+            default. This parameter is valid for Arm architectures only.
+
+        * ``initrd_image``: This points to the location of initrd image (e.g.,
+            ``rootfs.cpio.xz``) which will be used as target's root filesystem if kernel
+            does not include one already.
+
+        * ``mem_size``: Size of guest memory in MiB.
+
+        * ``num_cores``: Number of CPU cores. Guest will have ``2`` cores by default.
+
+        * ``num_threads``: Number of CPU threads. Set to ``2`` by defaults.
+
+        * ``cmdline``: Kernel command line parameter. It only specifies console device in
+            default (i.e., ``console=ttyAMA0``) which is valid for Arm architectures.
+            May be changed to ``ttyS0`` for x86 platforms.
+
+        * ``enable_kvm``: Specifies if KVM will be used as accelerator in QEMU or not.
+            Enabled by default if host architecture matches with target's for improving
+            QEMU performance.
+    :type qemu_settings: Dict
+
+    :param connection_settings: the dictionary to store connection settings
+        of :attr:`Target.connection_settings`, defaults to None.
+    :type connection_settings: Dict, optional
+
+    :param make_target: Lambda function for creating :class:`Target` based
+        object, defaults to :func:`lambda **kwargs: LinuxTarget(**kwargs)`.
+    :type make_target: func, optional
+
+    :Variable positional arguments: Forwarded to :class:`TargetRunner`.
+
+    :raises FileNotFoundError: if QEMU executable, kernel or initrd image cannot be found.
+    """
+
+    def __init__(self,
+                 qemu_settings,
+                 connection_settings=None,
+                 # pylint: disable=unnecessary-lambda
+                 make_target=lambda **kwargs: LinuxTarget(**kwargs),
+                 **args):
+        self.connection_settings = {
+            'host': '127.0.0.1',
+            'port': 8022,
+            'username': 'root',
+            'password': 'root',
+            'strict_host_check': False,
+        }
+
+        if connection_settings is not None:
+            self.connection_settings = self.connection_settings | connection_settings
+
+        qemu_args = {
+            'kernel_image': '',
+            'arch': 'aarch64',
+            'cpu_type': 'cortex-a72',
+            'initrd_image': '',
+            'mem_size': 512,
+            'num_cores': 2,
+            'num_threads': 2,
+            'cmdline': 'console=ttyAMA0',
+            'enable_kvm': True,
+        }
+
+        qemu_args = qemu_args | qemu_settings
+
+        qemu_executable = f'qemu-system-{qemu_args["arch"]}'
+        qemu_path = which(qemu_executable)
+        if qemu_path is None:
+            raise FileNotFoundError(f'Cannot find {qemu_executable} executable!')
+
+        if not os.path.exists(qemu_args["kernel_image"]):
+            raise FileNotFoundError(f'{qemu_args["kernel_image"]} does not exist!')
+
+        # pylint: disable=consider-using-f-string
```

Comment:
I'd agree with pylint here, that would make the code more readable. If that was not possible for whatever reason, I'd still use the keyword arguments of format() to provide named placeholders, rather than positional ones. Otherwise it's very easy to shift everything by one location and break the code, making it needlessly harder to modify in the future

Reply:
Changed it to f-string.